### DISCUSSION TITLE
Tuya 10 Gang Switch TS0601_switch_10

### DISCUSSION THREAD

**Discussion by Neil-M0NFI** [Koenkk/zigbee2mqtt#19938] [Devices]
I recently received a 10 gang Tuya switch from Aliexpress that wasn't supported 

![10_gang](https://github.com/Koenkk/zigbee2mqtt/assets/5224632/802a87b9-4671-449a-b6e2-b1eb79e7b278)

https://www.aliexpress.com/item/1005005847728198.html

I modified other similar switch files into this external converter file tze200_10.js file and will hopefully be useful to others until it is supported directly

`const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;
const tuya = require('zigbee-herdsman-converters/lib/tuya');

const definition = {
    // Since a lot of TuYa devices use the same modelID, but use different datapoints
    // it's necessary to provide a fingerprint instead of a zigbeeModel
    fingerprint: [
        {
            // The model ID from: Device with modelID 'TS0601' is not supported
            // You may need to add \u0000 at the end of the name in some cases
            modelID: 'TS0601',
            // The manufacturer name from: Device with modelID 'TS0601' is not supported.
            manufacturerName: '_TZE200_7sjncirf',
        },
    ],
    model: 'TS0601_switch_10',
    vendor: 'TuYa',
    description: '10 gang switch',
    fromZigbee: [tuya.fz.datapoints],
    toZigbee: [tuya.tz.datapoints],
//    onEvent: tuya.onEventSetTime, // Add this if you are getting no converter for 'commandMcuSyncTime'
    configure: tuya.configureMagicPacket,
    exposes: [
      e.switch().withEndpoint('l1'), 
      e.switch().withEndpoint('l2'), 
      e.switch().withEndpoint('l3'), 
      e.switch().withEndpoint('l4'), 
      e.switch().withEndpoint('l5'),
      e.switch().withEndpoint('l6'), 
      e.switch().withEndpoint('l7'), 
      e.switch().withEndpoint('l8'), 
      e.switch().withEndpoint('l9'), 
      e.switch().withEndpoint('l10')
  ],
  endpoint: (device) => {
      return {'l1': 1, 'l2': 1, 'l3': 1, 'l4': 1, 'l5': 1, 'l6': 1, 'l7': 1, 'l8': 1, 'l9': 1, 'l10': 1};
  },
  meta: {
      multiEndpoint: true,
      tuyaDatapoints: [
          [1, 'state_l1', tuya.valueConverter.onOff],
          [2, 'state_l2', tuya.valueConverter.onOff],
          [3, 'state_l3', tuya.valueConverter.onOff],
          [4, 'state_l4', tuya.valueConverter.onOff],
          [5, 'state_l5', tuya.valueConverter.onOff],
          [6, 'state_l6', tuya.valueConverter.onOff],
          [0x65, 'state_l7', tuya.valueConverter.onOff],
          [0x66, 'state_l8', tuya.valueConverter.onOff],
          [0x67, 'state_l9', tuya.valueConverter.onOff],
          [0x68, 'state_l10', tuya.valueConverter.onOff],
      ],
  },
};

module.exports = definition;`

**Comment by Koenkk** [comment]
Added!

Changes will be available in the dev branch in a few hours from now. (https://www.zigbee2mqtt.io/advanced/more/switch-to-dev-branch.html)

**Reply by Neil-M0NFI** [reply]
Thank you

**Comment by Exhibit99** [comment]
I have several of these switches and so installed the edge version but the device still says unsupported. 

What can I do to get this working in HA?  I am a novice and despite reading the external  converters section on the mqtt site I just do not understand what I am supposed to do.  Is it a case of copying your code above and pasting it somewhere?  If I try that in the dev section of the device there does not seem to be a save button.

**Comment by Exhibit99** [comment]
So using zigbee2mqtt edge and go to dev console and click generate I get this:



const definition = {
    zigbeeModel: ['TS0601'],
    model: 'TS0601',
    vendor: '_TZE204_7sjncirf',
    description: 'Automatically generated definition',
    extend: [],
    meta: {},
};

module.exports = definition;

What do I do next?

**Comment by Exhibit99** [comment]
Could the edge version be updated, or main version, to include the switch.  I have 9 of these switches with the item number _TZE204_7sjncirf

**Reply by Exhibit99** [reply]
any update as to how to integrate _TZE204_7sjncirf  I have installed the Edge version but it still says unsupported

**Reply by Neil-M0NFI** [reply]
Please see my post below, I've just tried anothe of these switches and like yours it has the manufacturer name: _TZE204_7sjncirf. I've got working with an external converter for now.

**Comment by Neil-M0NFI** [comment]
I've just tried another of these 10 gang switches, but as mentioned above by Exhibit99 it was unsupported because it has a different manufacturer's name.

manufacturerName: '_TZE204_7sjncirf'

I created an external converter identical to the one I created above with the new manufacturer name and const extend changed to const modernExtend, and it worked correctly.

Is it possible to add manufacturerName: '_TZE204_7sjncirf' to the supported switch with manufacturerName: '_TZE200_7sjncirf' so both types will be supported, please?



This is the external converter I tested and it works.

const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const modernExtend = require('zigbee-herdsman-converters/lib/modernExtend');
const e = exposes.presets;
const ea = exposes.access;
const tuya = require('zigbee-herdsman-converters/lib/tuya');

const definition = {
    // Since a lot of TuYa devices use the same modelID, but use different datapoints
    // it's necessary to provide a fingerprint instead of a zigbeeModel
    fingerprint: [
        {
            // The model ID from: Device with modelID 'TS0601' is not supported
            // You may need to add \u0000 at the end of the name in some cases
            modelID: 'TS0601',
            // The manufacturer name from: Device with modelID 'TS0601' is not supported.
            manufacturerName: '_TZE204_7sjncirf',
        },
    ],
    model: 'TS0601_switch_10',
    vendor: 'TuYa',
    description: '10 gang switch',
    fromZigbee: [tuya.fz.datapoints],
    toZigbee: [tuya.tz.datapoints],
//    onEvent: tuya.onEventSetTime, // Add this if you are getting no converter for 'commandMcuSyncTime'
    configure: tuya.configureMagicPacket,
    exposes: [
      e.switch().withEndpoint('l1'), 
      e.switch().withEndpoint('l2'), 
      e.switch().withEndpoint('l3'), 
      e.switch().withEndpoint('l4'), 
      e.switch().withEndpoint('l5'),
      e.switch().withEndpoint('l6'), 
      e.switch().withEndpoint('l7'), 
      e.switch().withEndpoint('l8'), 
      e.switch().withEndpoint('l9'), 
      e.switch().withEndpoint('l10')
  ],
  endpoint: (device) => {
      return {'l1': 1, 'l2': 1, 'l3': 1, 'l4': 1, 'l5': 1, 'l6': 1, 'l7': 1, 'l8': 1, 'l9': 1, 'l10': 1};
  },
  meta: {
      multiEndpoint: true,
      tuyaDatapoints: [
          [1, 'state_l1', tuya.valueConverter.onOff],
          [2, 'state_l2', tuya.valueConverter.onOff],
          [3, 'state_l3', tuya.valueConverter.onOff],
          [4, 'state_l4', tuya.valueConverter.onOff],
          [5, 'state_l5', tuya.valueConverter.onOff],
          [6, 'state_l6', tuya.valueConverter.onOff],
          [0x65, 'state_l7', tuya.valueConverter.onOff],
          [0x66, 'state_l8', tuya.valueConverter.onOff],
          [0x67, 'state_l9', tuya.valueConverter.onOff],
          [0x68, 'state_l10', tuya.valueConverter.onOff],
      ],
  },
};

module.exports = definition;

**Comment by Exhibit99** [comment]
thank you, after many hours with google AI I too got an external one created also

**Reply by Neil-M0NFI** [reply]
This new version of the 10 gang switch has been added and will be supported once the updates roll out

**Comment by Exhibit99** [comment]
That is great news. Thanks

Jonathan Turner
Signal: @JT_4BB.01
Whatsapp: 0203 995 0866

4 Breams Buildings
Clerks:020 7092 1900

***@***.***

Sent from [Proton Mail](https://proton.me/mail/home) for Android.

-------- Original Message --------
